### ISSUE TITLE
TmServer: `create` operations are not idempotent

### ISSUE THREAD

**Issue by riccardomurri** [TissueMAPS/TissueMAPS#89]
When creating a plate that already exists, the TmServer API returns an
HTTP 500 error (server-side error)::

        500 server error: internal server error for url: http://.../api/experiments/dg1hchmx/plates

Indeed, the server-side logs show the following error:

        [ 15730] 2017-10-30 10:48:28 | ERROR    | tmserver.appfactory                      | database integrity error: (psycopg2.IntegrityError) duplicate key value violates unique constraint "plates_name_key" DETAIL:  Key (name)=(plate1) already exists.

This is *not* a real error: it just means that the plate already
exists; the right thing to do is to intercept it and return an OK
status to the caller (named plate already exists).

Similar errors happen with "create acquisition" etc.

**Comment by riccardomurri** [comment]
I agree that one can conceive an API where "create" fails if the object to be created already exists -- this is indeed the case for SQL's `CREATE DATABASE` statements.  Similarly, one can object that one of the side-effects of "creation" is that the newly-created object is "clean slate", and that would be a supporting argument for making "create" fail if the target already exists.

My counter-argument is that (in the TM case), having an idempotent "create" (= no fail if target exists) is **more useful**.

Indeed, (re)creation of an experiment or plate or acquisition is something that is only done when uploading data to be processed.  If the upload process was interrupted, it makes ones' life simpler to just re-run the same upload process or script, without dealing with spurious errors.  (I would actually go as far as saying that experiments should be immutable and once the upload process is over, no modifications to the dataset structure should be allowed.)

**Comment by scottberry** [comment]
regarding "experiments should be immutable". On several occasions I have had to delete some of the files from the database that were inadvertently uploaded. This has saved me re-uploading the entire dataset. I would suggest that this behavior remain possible.

**Comment by riccardomurri** [comment]
> On several occasions I have had to delete some of the files from the database that were inadvertently uploaded. 

I was thinking of something like "when you start processing", the dataset is "freezed".
But anyway, not the main point here ;-)

**Comment by hackermd** [comment]
I agree that this is not ideal. The main problem is in my opinion that the HTTP status code is not informative, i.e. client programs can not handle the error, since they don't know why the request failed.

The reply for successful creation of a new resource should probably be 201 and 200 or 202 in case it already exists.
